oneflow/api/python/functional/py_function.h

```diff
@@ -25,31 +26,64 @@ namespace oneflow {
-inline py::object PyFunction(py::args args, py::kwargs kwargs) {
-  // TODO(): Support multiple function signatures.
-  CHECK_LE_OR_THROW(args.size(), SchemaT::max_positionals)
-      << "The maximum count of positional arguments is " << SchemaT::max_positionals;
-  CHECK_LE_OR_THROW(kwargs.size(), SchemaT::max_keywords)
-      << "The maximum count of keyword arguments is " << SchemaT::max_keywords;
-
-  // TODO(): Check argument types.
-  std::vector<PythonArg> _args(SchemaT::max_args);
-  for (int i = 0; i < args.size(); ++i) { _args[i] = PythonArg(args[i]); }
-  for (int i = args.size(); i < SchemaT::max_args; ++i) {
-    const auto& arg = SchemaT::argument_def.at(i);
-    if (kwargs.contains(arg.name.c_str())) {
-      _args[i] = PythonArg(kwargs[arg.name.c_str()]);
-    } else {
-      CHECK_OR_THROW(arg.has_default_value)
-          << "Argument " << arg.name << " is required, and the function def is \""
-          << SchemaT::signature << "\".";
-      _args[i] = PythonArg(arg.default_value);
+bool ParseArgs(const py::args& args, const py::kwargs& kwargs,  // NOLINT
+               std::vector<PythonArg>* parsed_args, const FunctionDef& function,
+               size_t max_pos_args, bool raise_exception);
+
+template<typename... SchemaT>
+class PyFunctionDispatcher {
+ public:
+  static_assert(sizeof...(SchemaT) >= 1, "Requires 1 template argument at least.");
+  using T0 = typename std::tuple_element<0, std::tuple<SchemaT...>>::type;
+
+  PyFunctionDispatcher() : schema_size_(sizeof...(SchemaT)), func_name_(T0::function_def.name) {
+    signatures_.resize(schema_size_);
+    RecursiveInit(std::make_index_sequence<sizeof...(SchemaT)>{});
+  }
+
+  template<size_t I0, size_t... I>
+  py::object call(const py::args& args, const py::kwargs& kwargs,
+                  std::index_sequence<I0, I...>) const {
+    using T = typename std::tuple_element<I0, std::tuple<SchemaT...>>::type;
+    std::vector<PythonArg> parsed_args(T::max_args);
+    if (ParseArgs(args, kwargs, &parsed_args, T::function_def, T::max_pos_args,
+                  /*raise_exception*/ schema_size_ == 1)) {
+      return detail::unpack_call(*T::func, parsed_args);
+    }
+    return call(args, kwargs, std::index_sequence<I...>{});
+  }
+
+  py::object call(const py::args& args, const py::kwargs& kwargs, std::index_sequence<>) const {
+    std::ostringstream ss;
+    ss << func_name_
+       << "(): received an invalid combination of arguments. The valid signatures are:";
+    for (int i = 0; i < signatures_.size(); ++i) {
+      ss << "\n\t*" << i << ": " << signatures_.at(i);
     }
+    THROW(TypeError) << ss.str();
+    return py::none();
+  }
+
+ private:
+  void RecursiveInit(std::index_sequence<>) {}
+
+  template<size_t I0, size_t... I>
+  void RecursiveInit(std::index_sequence<I0, I...>) {
+    using T = typename std::tuple_element<I0, std::tuple<SchemaT...>>::type;
+    signatures_[I0] = T::signature;
+    RecursiveInit(std::index_sequence<I...>{});
   }
```

Comment:
```suggestion
  template<size_t I>
  using get_sig_type = typename std::tuple_element<I, std::tuple<SchemaT...>>::type;
  template<size_t... I>
  void RecursiveInit(std::index_sequence<I...>) {
    __attribute__((__unused__)) int dummy = {((void)(signatures_[I] = get_sig_type<I, SchemaT...>::signature), 0)...};
  }
```

Reply:
the recursion can be eliminated before the compiler's optimization phase

Reply:
done